apiserver/facades/agent/controllercharm/controllercharm.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controllercharm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/environs/bootstrap"
+	"github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/state"
+)
+
+const (
+	// JujuMetricsUserPrefix defines the "namespace" in which this facade is
+	// allowed to create/remove users.
+	JujuMetricsUserPrefix = "juju-metrics-"
+
+	// UserCreator is the listed "creator" of metrics users in state.
+	UserCreator = "admin"
+)
+
+// ControllerCharmAPI provides API methods to the controllercharm worker.
+type ControllerCharmAPI struct {
+	state *state.State
+}
+
+// AddMetricsUser creates a user with the given username and password, and
+// grants the new user permission to read the metrics endpoint.
+func (api *ControllerCharmAPI) AddMetricsUser(args params.AddUsers) (params.AddUserResults, error) {
+	var results params.AddUserResults
+	for _, user := range args.Users {
+		err := api.addMetricsUser(user)
+		results.Results = append(results.Results, params.AddUserResult{
+			Tag:   user.Username,
+			Error: apiservererrors.ServerError(err),
+		})
+	}
+	return results, nil
+}
+
+func (api *ControllerCharmAPI) addMetricsUser(args params.AddUser) error {
+	if !strings.HasPrefix(args.Username, JujuMetricsUserPrefix) {
+		return fmt.Errorf("username %q should have prefix %q", args.Username, JujuMetricsUserPrefix)
+	}
+
+	if args.Password == "" {
+		return fmt.Errorf("must specify password for user %q", args.Username)
+	}
+
+	_, err := api.state.AddUser(args.Username, args.DisplayName, args.Password, UserCreator)
+	if err != nil {
+		return errors.Annotatef(err, "failed to create user %q", args.Username)
+	}
+
+	model, err := api.state.Model()
+	if err != nil {
+		return errors.Annotatef(err, "retrieving current model")
+	}
+
+	userTag := names.NewUserTag(args.Username)
+	_, err = model.AddUser(state.UserAccessSpec{
+		User:      userTag,
+		CreatedBy: names.NewUserTag(UserCreator),
+		Access:    permission.ReadAccess, // TODO: restrict access further?
```

Comment:
Currently, the user needs either `superuser` access on the controller or `read` access on the `controller` model to access the metrics endpoint. See
https://github.com/juju/juju/blob/ec839429ce0187c648080c7114c8d57e27423729/apiserver/introspection.go#L34
We're thinking about adding a new `read-metrics` permission which would restrict access even further, so that the user could *only* access the metrics endpoint (and not anything about deployed applications). I'm not sure if this is worth the squeeze though.